Include additional translation in getBBox for SVGUseElement

Ahmad-S792 opened the pull request:
#### e53a0bba5e47721f29a3e0770dcb207dcc4d712e
<pre>
Include additional translation in getBBox for SVGUseElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=264010">https://bugs.webkit.org/show_bug.cgi?id=264010</a>
<a href="https://rdar.apple.com/118082623">rdar://118082623</a>

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/99b52757a16d0a300b77c83c27c443d70ba78f9c">https://chromium.googlesource.com/chromium/blink/+/99b52757a16d0a300b77c83c27c443d70ba78f9c</a>

RenderObject::objectBoundingBox() can&apos;t be used, because the additional
translation is included in the local-to-parent transform.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGTransformableContainer.h:
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::getBBox):
* Source/WebCore/svg/SVGUseElement.h:
* LayoutTests/svg/custom/getBBox-use.html: Added.
* LayoutTests/svg/custom/getBBox-use-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGraphicsElement.getBBox-05-expected.txt: Progression

Canonical link: <a href="https://commits.webkit.org/302400@main">https://commits.webkit.org/302400@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c9056690fa2692a9cc7e1f3a9471e4ce1c50b76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128794 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39624 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136176 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80166 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d7edb4bc-ea37-463c-a3cf-44f53a0dfb78) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/928 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98050 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65971 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bfac4158-94e5-44c1-8bbd-76544f0d0171) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131741 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/754 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115407 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78659 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/98db4458-a138-4e23-925f-44f5a85987d0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/686 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33491 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79457 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109125 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33985 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138635 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/868 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/843 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106618 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111715 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106397 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27129 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/714 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30246 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53290 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/937 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64225 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/788 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/844 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/879 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on LayoutTests/svg/custom/getBBox-use.html:
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<script src="../../resources/testharness.js"></script>
+<script src="../../resources/testharnessreport.js"></script>
```
@Ahmad-S792 could we put this in WPT?

Ahmad-S792 replied:
Need to check with Chromium folks.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
```diff
@@ -40,6 +40,8 @@ class LegacyRenderSVGContainer : public LegacyRenderSVGModelObject {
     bool isRepaintSuspendedForChildren() const { return m_repaintIsSuspendedForChildrenDuringLayout; }
 
+    FloatRect objectBoundingBoxForUse() const { return objectBoundingBox(); }
+
```
Why do not we make objectBoundingBox() public instead of introducing a new function with a different name?

Ahmad-S792 replied:
I thought we didn't want it to be public so created this helper but now undid it.

shallawa commented on Source/WebCore/svg/SVGUseElement.cpp:
```diff
@@ -151,6 +151,29 @@ inline Document* SVGUseElement::externalDocument() const
+    // correct without additional work. Although it does not work due to
+    // problem stemming from including the additional translation directly
+    // on the renderer corresponding to the SVGUseElement.
+    FloatRect bbox = transformableContainer->objectBoundingBoxForUse();
```
Instead of all of this cannot we just call
```
auto bbox = SVGGraphicsElement::getBBox(styleUpdateStrategy);
```

Ahmad-S792 replied:
Not super clear here. Let me explore.

Ahmad-S792 replied:
@shallawa - Understood and now seems to be done.

shallawa commented on Source/WebCore/svg/SVGUseElement.h:
```diff
@@ -105,6 +108,7 @@ class SVGUseElement final : public SVGGraphicsElement, public SVGURIReference, p
     bool m_haveFiredLoadEvent { false };
     bool m_errorOccurred { false };
     bool m_shadowTreeNeedsUpdate { true };
+    Markable<FloatRect> m_objectBoundingBox;
```
I think this member is not used.

Ahmad-S792 replied:
Done. Was trying some thing and left it here.

shallawa commented on Source/WebCore/svg/SVGUseElement.cpp:
```diff
@@ -152,6 +152,19 @@ inline Document* SVGUseElement::externalDocument() const
+    auto bbox = SVGGraphicsElement::getBBox(styleUpdateStrategy);
+
+    auto* transformableContainer = dynamicDowncast<LegacyRenderSVGTransformableContainer>(renderer());
+    // Don't apply the additional translation if the oBB is invalid.
```
This comment does not add much. It just says what the code below does but in English.

Ahmad-S792 replied:
Removed.

shallawa commented on Source/WebCore/svg/SVGUseElement.cpp:
```diff
@@ -152,6 +152,19 @@ inline Document* SVGUseElement::externalDocument() const
 
+FloatRect SVGUseElement::getBBox(StyleUpdateStrategy styleUpdateStrategy)
+{
+    auto bbox = SVGGraphicsElement::getBBox(styleUpdateStrategy);
```
We should return `{ }` here `if (bbox.isEmpty())` is true.

Ahmad-S792 replied:
Done.

shallawa commented on Source/WebCore/svg/SVGUseElement.cpp:
```diff
@@ -152,6 +152,19 @@ inline Document* SVGUseElement::externalDocument() const
+    auto* transformableContainer = dynamicDowncast<LegacyRenderSVGTransformableContainer>(renderer());
+    // Don't apply the additional translation if the oBB is invalid.
+    if (!transformableContainer || !transformableContainer->isObjectBoundingBoxValid())
+        return { };
```
If we return above `if (bbox.isEmpty())` , then `renderer()` here has to be valid and it has to be of type `LegacyRenderSVGTransformableContainer`. So we can change this if-statement by an ASSERT instead.
```
ASSERT(transformableContainer && transformableContainer->isObjectBoundingBoxValid());
```

Ahmad-S792 replied:
Done.

Ahmad-S792 committed:
Include additional translation in getBBox for SVGUseElement
https://bugs.webkit.org/show_bug.cgi?id=264010
rdar://118082623

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/blink/+/99b52757a16d0a300b77c83c27c443d70ba78f9c

RenderObject::objectBoundingBox() can't be used, because the additional
translation is included in the local-to-parent transform.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGTransformableContainer.h:
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::getBBox):
* Source/WebCore/svg/SVGUseElement.h:
* LayoutTests/svg/custom/getBBox-use.html: Added.
* LayoutTests/svg/custom/getBBox-use-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGraphicsElement.getBBox-05-expected.txt: Progression

Canonical link: https://commits.webkit.org/302400@main
